File: rust/crates/belfast/src/helpers/axis_helper.rs
use crate::{
    BelfastError, BelfastResult, BindGroup, Buffer, BufferUsage, Device, Draw, DrawOptions, Mesh,
    VertexAttributeDescriptor, VertexBufferBinding,
};

pub struct AxisHelperOptions<'a> {
    pub label: &'a str,
    pub length: f32,
    pub format: wgpu::TextureFormat,
    pub layout: &'a wgpu::PipelineLayout,
    pub depth_stencil: Option<wgpu::DepthStencilState>,
}

impl<'a> AxisHelperOptions<'a> {
    pub fn new(format: wgpu::TextureFormat, layout: &'a wgpu::PipelineLayout) -> Self {
        Self {
            label: "AxisHelper",
            length: 1.0,
            format,
            layout,
            depth_stencil: None,
        }
    }
}

pub struct AxisHelper {
    mesh: Mesh,
    draw: Draw,
}

impl AxisHelper {
    pub fn new(device: &Device, options: AxisHelperOptions<'_>) -> BelfastResult<Self> {
        if !options.length.is_finite() || options.length <= 0.0 {
            return Err(BelfastError::InvalidAxisLength);
        }

        let (positions, colors) = axis_geometry(options.length);
        let position_buffer = Buffer::from_data(
            device,
            &positions,
            BufferUsage::vertex(),
            "AxisHelperPositions",
        );
        let color_buffer =
            Buffer::from_data(device, &colors, BufferUsage::vertex(), "AxisHelperColors");
        let mesh = Mesh::new(6)?
            .add_vertex_buffer(VertexBufferBinding {
                buffer: position_buffer,
                array_stride: 12,
                attributes: vec![VertexAttributeDescriptor {
                    shader_location: 0,
                    format: wgpu::VertexFormat::Float32x3,
                    offset: 0,
                }],
                slot: Some(0),
                step_mode: None,
            })?
            .add_vertex_buffer(VertexBufferBinding {
                buffer: color_buffer,
                array_stride: 12,
                attributes: vec![VertexAttributeDescriptor {
                    shader_location: 1,
                    format: wgpu::VertexFormat::Float32x3,
                    offset: 0,
                }],
                slot: Some(1),
                step_mode: None,
            })?;

        let mut draw_options = DrawOptions::new(options.label, options.format);
        draw_options.layout = Some(options.layout);
        draw_options.primitive.topology = wgpu::PrimitiveTopology::LineList;
        draw_options.depth_stencil = options.depth_stencil;
        let draw = Draw::new(
            device,
            include_str!("axis_helper.wgsl"),
            &mesh,
            draw_options,
        );

        Ok(Self { mesh, draw })
    }

    pub fn draw<'a>(&'a self, pass: &mut wgpu::RenderPass<'a>, camera_bind_group: &'a BindGroup) {
        camera_bind_group.bind(pass, 0);
        self.draw.draw(pass, &self.mesh, 1);
    }
}

fn axis_geometry(length: f32) -> ([f32; 18], [f32; 18]) {
    (
        [
            0.0, 0.0, 0.0, length, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, length, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, length,
        ],
        [
            1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0,
            1.0,
        ],
    )
}

#[cfg(test)]
mod tests {
    use super::axis_geometry;

    #[test]
    fn geometry_contains_three_colored_positive_axes() {
        let (positions, colors) = axis_geometry(2.0);
        assert_eq!(&positions[3..6], &[2.0, 0.0, 0.0]);
        assert_eq!(&positions[9..12], &[0.0, 2.0, 0.0]);
        assert_eq!(&positions[15..18], &[0.0, 0.0, 2.0]);
        assert_eq!(&colors[0..6], &[1.0, 0.0, 0.0, 1.0, 0.0, 0.0]);
        assert_eq!(&colors[12..18], &[0.0, 0.0, 1.0, 0.0, 0.0, 1.0]);
    }
}

File: archive/packages/belfast/src/helper/AxisHelper.ts
import type { BindGroup } from "../core/BindGroup";
import type { Device } from "../core/Device";
import { Buffer, BufferUsage } from "../core/Buffer";
import { Mesh } from "../core/Mesh";
import { Draw } from "./Draw";
import { createSceneUniformPipelineLayout } from "./sceneLayout";

const AXIS_SHADER = /* wgsl */ `
struct SceneUniforms {
  viewProj: mat4x4<f32>,
}

@group(0) @binding(0) var<uniform> scene: SceneUniforms;

struct VertexInput {
  @location(0) position: vec3<f32>,
  @location(1) color: vec3<f32>,
}

struct VertexOutput {
  @builtin(position) position: vec4<f32>,
  @location(0) color: vec3<f32>,
}

@vertex
fn vs_main(input: VertexInput) -> VertexOutput {
  var output: VertexOutput;
  output.position = scene.viewProj * vec4<f32>(input.position, 1.0);
  output.color = input.color;
  return output;
}

@fragment
fn fs_main(input: VertexOutput) -> @location(0) vec4<f32> {
  return vec4<f32>(input.color, 1.0);
}
`;

const DEFAULT_LENGTH = 1000;

export interface AxisHelperOptions {
  /** Half-extent of each axis line (default 1000, matching alfrid DrawAxis). */
  length?: number;
  label?: string;
  /**
   * Shared with other scene draws (e.g. triangle `Draw`) so one bind group works.
   * From `createSceneUniformPipelineLayout(device).pipelineLayout`.
   */
  pipelineLayout?: GPUPipelineLayout;
}

function buildAxisGeometry(length: number): {
  positions: Float32Array;
  colors: Float32Array;
} {
  const positions = new Float32Array([
    -length,
    0,
    0,
    length,
    0,
    0,
    0,
    -length,
    0,
    0,
    length,
    0,
    0,
    0,
    -length,
    0,
    0,
    length,
  ]);
  const colors = new Float32Array([1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1]);
  return { positions, colors };
}

/**
 * RGB axis lines along X (red), Y (green), and Z (blue).
 * Uses `line-list` topology; share a view-projection bind group with other scene draws.
 */
export class AxisHelper {
  readonly mesh: Mesh;
  private readonly lineDraw: Draw;
  private readonly positionBuffer: Buffer;
  private readonly colorBuffer: Buffer;

  constructor(device: Device, options: AxisHelperOptions = {}) {
    const length = options.length ?? DEFAULT_LENGTH;
    const label = options.label ?? "AxisHelper";
    const { vertex } = BufferUsage;
    const { positions, colors } = buildAxisGeometry(length);

    this.positionBuffer = Buffer.fromData(device, positions, vertex, `${label}-positions`);
    this.colorBuffer = Buffer.fromData(device, colors, vertex, `${label}-colors`);

    this.mesh = new Mesh(6)
      .addVertexBuffer({
        buffer: this.positionBuffer,
        arrayStride: 12,
        attributes: [{ shaderLocation: 0, format: "float32x3", offset: 0 }],
        slot: 0,
      })
      .addVertexBuffer({
        buffer: this.colorBuffer,
        arrayStride: 12,
        attributes: [{ shaderLocation: 1, format: "float32x3", offset: 0 }],
        slot: 1,
      });

    const pipelineLayout =
      options.pipelineLayout ??
      createSceneUniformPipelineLayout(device, `${label}Scene`).pipelineLayout;

    this.lineDraw = new Draw(device, AXIS_SHADER, {
      label,
      layout: pipelineLayout,
      primitive: { topology: "line-list" },
      vertexBuffers: this.mesh.getVertexLayouts(),
      depthStencil: {
        format: "depth24plus",
        depthWriteEnabled: true,
        depthCompare: "less",
      },
    });
  }

  /** Returns bind group layout index 0 (same `SceneUniforms` as other camera-lit helpers). */
  getBindGroupLayout(index = 0): GPUBindGroupLayout {
    return this.lineDraw.getBindGroupLayout(index);
  }

  draw(passEncoder: GPURenderPassEncoder, bindGroup: BindGroup): void {
    this.lineDraw.draw(passEncoder, this.mesh, bindGroup);
  }

  destroy(): void {
    this.positionBuffer.destroy();
    this.colorBuffer.destroy();
  }
}
